### PR TITLE
Still issues with unicode file names

### PR DESCRIPTION
**Pull request by mih**
#### What steps will reproduce the problem?
```
% datalad --dbg add .
                                                                                                                      Traceback (most recent call last):
  File "/home/mih/hacking/datalad/git/bin/datalad", line 8, in <module>
    main()
  File "/home/mih/hacking/datalad/git/datalad/cmdline/main.py", line 347, in main
    ret = cmdlineargs.func(cmdlineargs)
  File "/home/mih/hacking/datalad/git/datalad/interface/base.py", line 427, in call_from_parser
    ret = list(ret)
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 382, in generator_func
    result_renderer, result_xfm, _result_filter, **_kwargs):
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 449, in _process_results
    for res in results:
  File "/home/mih/hacking/datalad/git/datalad/distribution/add.py", line 424, in __call__
    **add_kw
  File "/home/mih/hacking/datalad/git/datalad/support/gitrepo.py", line 300, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/support/annexrepo.py", line 1522, in add
    expect_stderr=True
  File "/home/mih/hacking/datalad/git/datalad/support/annexrepo.py", line 2259, in _run_annex_command_json
    **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/support/annexrepo.py", line 1029, in _run_annex_command
    return self.cmd_call_wrapper.run(cmd_list, **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/cmd.py", line 588, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/cmd.py", line 428, in run
    expect_fail=expect_fail)
  File "/home/mih/hacking/datalad/git/datalad/cmd.py", line 245, in _get_output_online
    line = log_stdout_(line.decode())
UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 139: ordinal not in range(128)
()
> /home/mih/hacking/datalad/git/datalad/cmd.py(245)_get_output_online()
-> line = log_stdout_(line.decode())
(Pdb) line
'{"command":"add","success":true,"key":"MD5E-s19398299--8216274e520a345ac0fd51eed8564e99.pdf","file":"REDACTED_RELEVANTPART:Zo\xef\xbf\xbd\xef\xbf\xbd..."}\n'
```

### CODE REVIEW

**Comment by mih** [comment]
Just FTR: I have triangulated (some aspect of) this issue be happening in (Git)Runner... it is a mess, seriously.

**Comment by mih** [comment]
Here is a symptom: `Runner._process_one_line()` calls `stdout.readline()` which gives something like this:

```
>>>s=b'{"command":"add","note":"non-large file; adding content to git repository","success":true,"file":"\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd"}\n'
>>> s
b'{"command":"add","note":"non-large file; adding content to git repository","success":true,"file":"\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd\xef\xbf\xbd"}\n'
>>> s.decode()
'{"command":"add","note":"non-large file; adding content to git repository","success":true,"file":"���������"}\n'
```

which is the response to the runner executing:

```
['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'add', '--json', '-c', 'annex.largefiles=nothing', '--include-dotfiles', '--', '.', 'äöü東']
```

and this is under PY3, and we have dedicated PY3-code that executes `decode()`

I have no clue....

**Comment by kyleam** [comment]
I think it's happening outside of datalad at the level of what the --json output.  If I run

```
$ DATALAD_TESTS_TEMP_KEEP=1 nosetests -sv datalad/distribution/test/test_add.py:test_add_dirty_tree
```

Then go to that subdirectory and run manually run the annex add command, I see

```
$ git reset --mixed HEAD~2
$ git -c receive.autogc=0 -c gc.auto=0 annex add --json -- äöü東
{"command":"add","success":true,"key":"MD5E-s4--334c4a4c42fdb79d7ebc3e73b517e6f8","file":"���������"}

$ git status
On branch master
Changes to be committed:
        new file:   "\303\244\303\266\303\274\346\235\261"

Untracked files:
        testindir
```

**Comment by bpoldrack** [comment]
No, `readline()` delivers it correctly. It's just git that shows the wrong thing.

FTR: I could reproduce it and have a proof-of-concept fix:

explictly adding `'utf-8'` to the `encode()` and `decode()` calls in that `_process_one_line`. And then add `encode('utf-8')` to the calls to `res['path']` within `result_renderer``when formating the string to be passed into `ui.message`:

```
/tmp/test-unicode % datalad add .        
add(ok): /tmp/test-unicode/äöü東 (file)   
add(ok): /tmp/test-unicode (dataset)
save(ok): /tmp/test-unicode (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
```

Still not a real solution, but I'm close, I guess.

**Comment by bpoldrack** [comment]
@kyleam : My " it's just git" was a bit too fast for I started looking into it the same way. But I now notice, that you get indeed something different, than I do. So at least for you, there's an additional issue going on. I get the same for `git-status`, but not for the annex call:

```

/tmp/test-unicode % git -c receive.autogc=0 -c gc.auto=0 annex add --json -- äöü東
{"command":"add","success":true,"key":"SHA256E-s0--e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","file":"äöü東"}
```

**Comment by kyleam** [comment]
Hmm, I think the `git status` output (which I wasn't viewing as related to this issue) is just a matter of 

```
git -c core.quotepath=false status
On branch master
Changes to be committed:
	new file:   äöü東

Untracked files:
	testindir

```

**Comment by yarikoptic** [comment]
I am the (2nd) winner!!!
```shell
$> git annex add --json äöü東  
{"command":"add","success":true,"key":"SHA256E-s3--ca6692e1ef9d42f56dacaeaaa59fac1fa6f6475629ade61bd3fcf6a0cfdba1c2","file":"äöü東"}
```

**Comment by mih** [comment]
Awesome. There is so much winning here, we could close the issue. If only the tests would still fail ;-)

**Comment by yarikoptic** [comment]
cut/pasteable example to reproduce/compare details:
```shell
$> ( d=/tmp/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; git annex add --json "$f"; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )
Initialized empty Git repository in /tmp/test/.git/
init  ok
(recording state in git...)
{"command":"add","success":true,"key":"SHA256E-s2--4355a46b19d348dc2f57c046f8ef63d4538ebb936000f3c9ee954a27460dd865","file":"äöü東"}
LANG=en_US.utf8
LANGUAGE=
LC_CTYPE="en_US.utf8"
LC_NUMERIC="en_US.utf8"
LC_TIME="en_US.utf8"
LC_COLLATE="en_US.utf8"
LC_MONETARY="en_US.utf8"
LC_MESSAGES="en_US.utf8"
LC_PAPER="en_US.utf8"
LC_NAME="en_US.utf8"
LC_ADDRESS="en_US.utf8"
LC_TELEPHONE="en_US.utf8"
LC_MEASUREMENT="en_US.utf8"
LC_IDENTIFICATION="en_US.utf8"
LC_ALL=
git-annex version: 6.20180329+gitga5fe62bed-1~ndall+1
git version 2.16.2
utf-8
```

**Comment by yarikoptic** [comment]
@joeyh - any clue why we are getting different output from `git annex add --json` on a unicode filename?  the same git annex, git and locale seems to us.
Cheers!

**Comment by kyleam** [comment]
And presumably as a consequence

```
$ nosetests -sv datalad/distribution/tests/test_add.py:test_add_dirty_tree
```

passes for @yarikoptic on py3 but fails for me.

----

Here is what I see for with his snippet (with an additional echo).  It has the same problem with the file name as my `git annex --json add` output above.


```
$ ( d=/home/kyle/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; echo creating "$f"; git annex add --json "$f" >/tmp/out; echo "$f" >>/tmp/out; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )
Initialized empty Git repository in /home/kyle/test/.git/
init  ok
(recording state in git...)
creating äöü東
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
git-annex version: 6.20180329+gitga5fe62bed-1~ndall+1
git version 2.17.0
utf-8
```

**Comment by kyleam** [comment]
Sorry, I pasted the wrong snippet (it redirected annex add output into a file).  Here's the correct one:

```
$ ( d=/home/kyle/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; echo creating "$f"; git annex add --json "$f"; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )
Initialized empty Git repository in /home/kyle/test/.git/
init  ok
(recording state in git...)
creating äöü東
{"command":"add","success":true,"key":"SHA256E-s2--4355a46b19d348dc2f57c046f8ef63d4538ebb936000f3c9ee954a27460dd865","file":"���������"}
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
git-annex version: 6.20180329+gitga5fe62bed-1~ndall+1
git version 2.17.0
utf-8
```

**Comment by bpoldrack** [comment]
FTR: In current state `test_add.py` passes for me locally in PY2 and PY3. Didn't figure out yet, what's different on Travis - I'm still on it.

And for comparison @yarikoptic 'S test:
```
/tmp % ( d=/tmp/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; git annex add --json "$f"; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )
Initialized empty Git repository in /tmp/test/.git/
init  
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <(null)>) not allowed
ok
(recording state in git...)
{"command":"add","success":true,"key":"SHA256E-s2--4355a46b19d348dc2f57c046f8ef63d4538ebb936000f3c9ee954a27460dd865","file":"äöü東"}
LANG=en_US.utf8
LANGUAGE=
LC_CTYPE="en_US.utf8"
LC_NUMERIC="en_US.utf8"
LC_TIME="en_US.utf8"
LC_COLLATE="en_US.utf8"
LC_MONETARY="en_US.utf8"
LC_MESSAGES="en_US.utf8"
LC_PAPER="en_US.utf8"
LC_NAME="en_US.utf8"
LC_ADDRESS="en_US.utf8"
LC_TELEPHONE="en_US.utf8"
LC_MEASUREMENT="en_US.utf8"
LC_IDENTIFICATION="en_US.utf8"
LC_ALL=
git-annex version: 6.20180329+gitga5fe62bed-1~ndall+1
git version 2.13.0.rc1.294.g07d810a77f
utf-8
```

**Comment by kyleam** [comment]
@bpoldrack
> In current state test_add.py passes for me locally in PY2 and PY3

Did these tests pass for you on py3 pass before any of your changes on this branch (i.e., at 002c775b71f6f7963e88bdaac7d583760ad7e04b)?  I'm asking because 1) I believe @yarikoptic's py3 tests did pass on that commit and 2) I'm wondering if the py3 failures that @mih reports are due to the `git annex add --json` output that I describe.  As you say above, there seems like there might be two separate issues going on here.

**Comment by bpoldrack** [comment]
Hehe - you're right, @kyleam . PY3 passes for me at https://github.com/datalad/datalad/commit/002c775b71f6f7963e88bdaac7d583760ad7e04b. I didn't even notice. Probably because I was amidst fixing PY2 when I started to check PY3.

**Comment by mih** [comment]
```
% ( d=/tmp/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; git annex add --json "$f"; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )
Initialized empty Git repository in /tmp/test/.git/
init  ok
(recording state in git...)
{"command":"add","success":true,"key":"SHA256E-s2--4355a46b19d348dc2f57c046f8ef63d4538ebb936000f3c9ee954a27460dd865","file":"���������"}
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
git-annex version: 6.20180316+gitg308f3ecf6-1~ndall+1
git version 2.16.1
utf-8
```

**Comment by bpoldrack** [comment]
Ok, so it's pretty clear, that there's indeed an additional issue. @kyleam and @mih (and prob. Travis) get the "wrong thing" from `git-annex` itself. And it's neither about git/git-annex version nor the locale settings.

WTH?

**Comment by mih** [comment]
20 bucks on earth rays....

**Comment by bpoldrack** [comment]
Considering that this probably is the first PR ever, where OSX is the only buildbot passing instead of the only one failing, I'm all in on the earth rays.

**Comment by kyleam** [comment]
Joey Hess <notifications@github.com> writes:

> It varies based on LANG. With LANG=C 
>
> "file":"���������"
>
> With a utf-8 LANG
>
> "file":"äöü東"

LANG=C can trigger it, but I think there's more to it.  @mih and I both
have LANG=en_US.UTF-8, but we're seeing "file":"���������".

**Comment by mih** [comment]
Hmm, I always treated the `LANGUAGE` variable as intended to selected the _language_ of message, rather than the locale. The GNU manual has nothing on locales in the docs on `LANGUAGE`: https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html

If I override it, it doesn't seem to fix the issue:

```
% (LANGUAGE=en_US.UTF-8  d=/tmp/test; f='äöü東'; [ -e "$d" ] && chmod a+w -R $d; rm -rf $d; mkdir -p $d; cd $d; export HOME=$d; git init; git annex init; echo 1 > "$f"; git annex add --json "$f"; locale; git annex version 2>&1 | head -n 1 ; git --version; python3 -c 'import sys; print(sys.getfilesystemencoding())' )


Initialized empty Git repository in /tmp/test/.git/
init  ok
(recording state in git...)
{"command":"add","success":true,"key":"SHA256E-s2--4355a46b19d348dc2f57c046f8ef63d4538ebb936000f3c9ee954a27460dd865","file":"���������"}
LANG=en_US.UTF-8
LANGUAGE=en_US.UTF-8
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
git-annex version: 6.20180316+gitg308f3ecf6-1~ndall+1
git version 2.16.1
utf-8
```

**Comment by joeyh** [comment]
I don't know why your system might not behave as if it's in a utf-8
locale, and it seems like a very deep rabbit hole to investigate that,
since there are all kinds of ways to break a system as complicated as
the locale system.

In any case since it's easy to reproduce the bug with LANG=C,
I am already well on my way to fixing it in git-annex.

-- 
see shy jo

**Comment by joeyh** [comment]
fixed in git-annex 89e1a05a8f62ac6ac3ae3f24be0165fe1fcc6f75

-- 
see shy jo

**Comment by kyleam** [comment]
> fixed in git-annex 89e1a05a8f62ac6ac3ae3f24be0165fe1fcc6f75

That resolves the issue on my end.  Thanks.

**Comment by mih** [comment]
Cool, thanks. I guess we can merge the added tests in here as soon as the test beds get the fresh annex. Question is what we should do about @bpoldrack's modifications. Are they needed? Or improve things? Or should we revert them?

**Comment by kyleam** [comment]
> Question is what we should do about @bpoldrack's modifications. Are they needed? Or improve things? Or should we revert them?

They are needed.  The below test passes on c86bdc4b4, but fails before his commits:

```
002c775b7 py2> nosetests datalad/distribution/tests/test_add.py:test_add_dirty_tree
E
======================================================================
ERROR: datalad.distribution.tests.test_add.test_add_dirty_tree
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kyle/src/python/datalad/venv-datalad-py2/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/kyle/src/python/datalad/datalad/tests/utils.py", line 494, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/kyle/src/python/datalad/datalad/distribution/tests/test_add.py", line 182, in test_add_dirty_tree
    added2 = ds.add('dir/.', to_git=True)
  File "/home/kyle/src/python/datalad/datalad/distribution/dataset.py", line 439, in apply_func
    return f(**kwargs)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 470, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 458, in return_func
    results = list(results)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 414, in generator_func
    result_renderer, result_xfm, _result_filter, **_kwargs):
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 482, in _process_results
    for res in results:
  File "/home/kyle/src/python/datalad/datalad/distribution/add.py", line 423, in __call__
    **add_kw
  File "/home/kyle/src/python/datalad/datalad/support/gitrepo.py", line 300, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/kyle/src/python/datalad/datalad/support/annexrepo.py", line 1541, in add
    expect_stderr=True
  File "/home/kyle/src/python/datalad/datalad/support/annexrepo.py", line 2303, in _run_annex_command_json
    **kwargs)
  File "/home/kyle/src/python/datalad/datalad/support/annexrepo.py", line 1041, in _run_annex_command
    return self.cmd_call_wrapper.run(cmd_list, **kwargs)
  File "/home/kyle/src/python/datalad/datalad/cmd.py", line 665, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/kyle/src/python/datalad/datalad/cmd.py", line 495, in run
    expect_fail=expect_fail)
  File "/home/kyle/src/python/datalad/datalad/cmd.py", line 300, in _get_output_online
    stdout += self._process_one_line(*stdout_args)
  File "/home/kyle/src/python/datalad/datalad/cmd.py", line 346, in _process_one_line
    line = line.encode()
UnicodeEncodeError: 'ascii' codec can't encode characters in position 98-101: ordinal not in range(128)
```

**Comment by mih** [comment]
git-annex version 6.20180416+gitg86b18966f-1~ndall+1 resolves this for me. Thanks!

**Comment by mih** [comment]
Restarted travis to see how the new annex version does.

**Comment by mih** [comment]
Seems to be doing good, merged master, off to the build bots...

**Comment by mih** [comment]
But the build bots are not having the needed git-annex update...

**Comment by yarikoptic** [comment]
Are we requiring new annex version now in general? Then we should note so in AnnexRepo and I will upgrade it on bots. If not, then tests should tollerate older supported ones

**Comment by yarikoptic** [comment]
My point is just to not have tests failing when we start building releases

**Comment by mih** [comment]
Sure, if you want to be on the safe side tighten the deps for those packages. But in general we have evidence in this PR that it worked perfectly well on 50% of the tested systems with an older version -- we just do not know perfectly well why, hence cannot formulate appropriate conditions for a more flexible test.

**Comment by yarikoptic** [comment]
uff... this one is not fun, but what i really don't get is why on some buildbots it is only test_add.py which is having such issues - I would expect many other tests triggering problems in annotate paths functionality, but it solely in test_add.py somehow, e.g. see http://smaug.datalad.org:8020/builders/datalad-pr-docker-dl-nd14_04/builds/4144/steps/nosetests/logs/stdio  . On travis it is test_archives, for which I did fix things up locally.
Any ideas are welcome.... I am even ready to postpone finishing up fixing it until the release ;-)

**Comment by mih** [comment]
Looks all sane to me. But I don't have any ideas to add, unfortunately.

**Comment by yarikoptic** [comment]
ok, merged master and pushed

in above commits I have gave up and cheated -- we will not test on unicode if filesystem is reported to be the one which doesn't support it. At least locally it should be all unicode I hope ;)

**Comment by mih** [comment]
That is not giving up -- that is predictive optimization!

File: datalad/__init__.py [hunk 1]
```diff
@@ -160,6 +160,26 @@ def setup_package():
...
     import os, sys
+    from datalad.tests.utils import rmtemp, OBSCURE_FILENAME
+
+    lgr.debug("Printing versioning information collected so far")
+    from datalad.support.external_versions import external_versions as ev
+    print(ev.dumps(query=True))
+    print("Obscure filename: str=%s repr=%r"
+            % (OBSCURE_FILENAME.encode('utf-8'), OBSCURE_FILENAME))
+    print("Encodings: default=%r filesystem=%r"
+            % (sys.getdefaultencoding(), sys.getfilesystemencoding()))
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply datalad/__init__.py]: good idea! will do (but will not push right away -- want to see what those beasts report about those Ialready added).  I think it is also worth RFing a bit so the same information could be displayed in WTF

[the remark this answers is not in the document]
Comment (kyleam) [reply datalad/__init__.py]: > I think it is also worth RFing a bit so the same information could be displayed in WTF

yep, that seems useful

**Comment by yarikoptic** [comment]
uff... here Mac for you -- fancy enough even to rewrite (recode?) filename so it looks the same but it is different underneath and also failing tests on OSX... 
```shell

*>>> print(u' "\';b&b&c\u0394\u0418\u0306\u05e7\u0645\u0e57\u3042 `')
 "';b&b&cΔЙקم๗あ `
*>>> print(u' "\';b&b&c\u0394\u0419\u05e7\u0645\u0e57\u3042 ')
 "';b&b&cΔЙקم๗あ 
*>>> print(len(u' "\';b&b&c\u0394\u0419\u05e7\u0645\u0e57\u3042 '))
16
*>>> print(len(u' "\';b&b&c\u0394\u0418\u0306\u05e7\u0645\u0e57\u3042 `'))
18

*>>> print(u'\u0419')
Й
*>>> print(u'\u0418')
И
*>>> print(u'\u0418\u0306')
Й
```
so the last and 2 behind - look like the same letter composed differently though

**Comment by yarikoptic** [comment]
ok, on buildbots (to treat OSX) I now set LANG="en_US.UTF-8" for all of them that workarounds for failures on OSX.

**Comment by yarikoptic** [comment]
if tests pass, please remove TEMP commit and force push

**Comment by yarikoptic** [comment]
for all rebase lovers rebased the whole thing on master since spent some time trying to find that bloody assertion which was failing in the code on travis ;)

**Comment by yarikoptic** [comment]
>  how and when assure_bytes is meant to be called
In the best traditions of unicode handling in Python - when things do not work otherwise ;-)  I think it is just that Python assumes non unicode capable filesystem at some points where we feed it with our unicode filenames for os.path operations.  So largely this is to used in the points of invocation of those and thus there is the `support.path` now which would do it upon failure... crude and unelegant, I know

but I wonder also why we have
```
 /bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
git-annex: external special remote protocol error, unexpectedly received "" (unable to parse command)
```
reported while running that failing test (only)... 

PS will be driving for the rest of today so if you have a chance/idea -- feel free.  Do not merge as is though - needs some TEMP commits be poped

**Comment by mih** [comment]
Looks good. The test that fails on travis passes locally for me:

```
% DATALAD_REPO_DIRECT=yes WRAPT_DISABLE_EXTENSIONS=1 DATALAD_LOG_LEVEL=info DATALAD_TESTS_SSH=1 python -m nose -s datalad/customremotes/tests/test_archives.py:test_basic_scenario
[WARNING] Running drop resulted in stderr output: git-annex: drop: 1 failed
 
[INFO   ] To obtain some keys we need to fetch an archive of size 203 Bytes 
[WARNING] Running drop resulted in stderr output: git-annex: drop: 1 failed
 
[INFO   ] To obtain some keys we need to fetch an archive of size 204 Bytes 
.Versions: appdirs=1.4.3 boto=2.44.0 cmd:annex=6.20180522+gitg71cbc1012-1~ndall+1 cmd:git=2.17.0 cmd:system-git=2.16.1 cmd:system-ssh=7.6p1 exifread=2.1.2 git=2.1.8 gitdb=2.0.2 humanize=0.5.1 iso8601=0.1.11 msgpack=0.5.6 mutagen=1.38 patoolib=1.12 requests=2.18.1 scrapy=1.5.0 six=1.11.0 tqdm=4.20.0 wrapt=1.10.11
Obscure filename: str=b' "\';a&b&c\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 `| ' repr=' "\';a&b&cΔЙקم๗あ `| '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: LANG='en_US.UTF-8' LANGUAGE='en_US:en' PATH='/home/mih/env/datalad3-dev/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/home/mih/bin:/home/mih/bin/pycharm/bin' GIT_EDITOR='vim' LC_CTYPE='en_US.UTF-8' GIT_PYTHON_GIT_EXECUTABLE='/usr/lib/git-annex.linux/git'

----------------------------------------------------------------------
Ran 1 test in 12.398s

OK
```

All using the new secret annex build.

**Comment by yarikoptic** [comment]
passing locally is a no brainer ;) I guess now I will try with that yet another custom build ... current failure is
```
datalad.customremotes.tests.test_archives.test_basic_scenario ... /bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
  external special remote error: Problem processing 'CHECKURL' with parameters ['dl+archive:SHA256E-s206--5cd6ce9d7336b60714a552186be0b0121eb7c8607fd48b0182f4db3676dd8267.tar.gz#path=+%22%27%3Bb%26b%26c%CE%94%D0%99%D7%A7%D9%85%E0%B9%97%E3%81%82+%60%7C+/d/+%22%27%3Ba%26b%26c%CE%94%D0%99%D7%A7%D9%85%E0%B9%97%E3%81%82+%60%7C+']: "'ascii' codec can't decode byte 0xce in position 11: ordinal not in range(128) [ascii.py:decode:26]"
git-annex: addurl: 1 failed
```

**Comment by yarikoptic** [comment]
no way!  tests passed !  so... if git-annex wasn't setting LOCPATH we would be in clear.  Setting it even to the "correct" (?) path seems to cause python to fall back to ascii encoding causing us trouble...  I wonder if may be I should just patch out LOCPATH setting for NeuroDebian builds, what do you think @joeyh ?

**Comment by yarikoptic** [comment]
Thank you @joeyh for the details!  I have glanced over our old thread troubleshooting all that locale-related madness:
- standalone build was ran/tested on CentOS
- the main problem was the "fatal: Out of memory, getdelim failed" msg, which was a genuine (unrelated) bug in git and was fixed later on... and indeed setting LOCPATH=/dev/null somehow, in one of the usecases did manage to avoid hitting it (but did not entirely eliminate it)

So, I think that for now I will go the path of least resistance and just patch away LOCPATH setting on  the standalone build for neurodebian.  I will send a patch for your consideration.

Cheers!

**Comment by yarikoptic-private** [comment]
sweet -- tests pass with bleeding edge annex (on travis and locally) built with no LOCPATH patching, so I am uploading that one to neurodebian now, and will aim to finalize this PR

**Comment by mih** [comment]
This could be the end of an era!

**Comment by yarikoptic** [comment]
@joeyh I wonder if standalone build patching of LOCPATH also relates to https://github.com/nipy/heudiconv/issues/228 and my quick check on our cluster, where it is a sysadmin who "installs" (well -- just extracts) the standalone build somewhere to be used by regular users on the system.  Then if I use `git` from the standalone I do get
```
[yhalchen@discovery datalad]$ LC_ALL=en_CA.ISO-8859-1 git status                                   
/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_CA.ISO-8859-1)
...
```
where system wide works fine

**Comment by joeyh** [comment]
Comprehensively fixed in e802323071c10ef17790a820e9d5ba7b0ade686e

-- 
see shy jo

**Comment by joeyh** [comment]
I've made it fall back to a temporary directory if HOME is not
available.

(Almost out of patience on this though. There's only so many hacks on
top of hacks on top of hacks to work around things I don't want people
to do that I can take.)

-- 
see shy jo

**Comment by yarikoptic** [comment]
so it would regenerate in a new temporary directory upon every invocation... well -- my point is that I am not sure if those hacks are needed for neurodebian standalone builds.  IIRC our initial problems we ran into were so vague and git at that time was also "getdelim"-bug prone (#1678) that I am no longer sure if all the locales issue is there and that it is somehow likely to run into "incompatible locale files" or we could just assume/rely that having system wide locales installed is sufficient (for neurodebian pkg at least)

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
> so it would regenerate in a new temporary directory upon every invocation...
> well -- my point is that I am not sure if those hacks are needed for
> neurodebian standalone builds. IIRC our initial problems we ran into were so
> vague and git at that time was also "getdelim"-bug prone (#1678) that I am no
> longer sure if all the locales issue is there and that it is somehow likely to
> run into "incompatible locale files" or we could just assume/rely that having
> system wide locales installed is sufficient (for neurodebian pkg at least)

If you're not worried about glibc version skew, you should not be using
the standalone bundle. 

I'm out of patience with this, sorry.

-- 
see shy jo

**Comment by joeyh** [comment]
The getdelim crash was confirmed to be prevented by LOCPATH=/dev/null
(in message-id 20161004160549.GV9830@onerussian.com).
Therefore, we *do* know that glibc/locales version skew can cause
failures. Furthermore, things don't seem to check for those
failures -- in the getdelim case errno was set earlier by something
locale-related failing, and only happened to be exposed by the getdelim
code.

Furthermore, before the getdelim crash, there was a ghc bug, triggered
by glibc failing to find the locales files it expected to, which
led to git-annex hanging in a busy loop as the exception handler repeatedly
crashed while attempting to display an error message.
https://ghc.haskell.org/trac/ghc/ticket/7695

So, returning to glibc/locales version skew means that anything could
break at any time in some inexplocable way. I would not be surprised if
any number of unexplained segfaults or other weird behavior from before
that change was also caused by similar version skew.

Faced with a choice between supporting 100% of things to an uncertain
degree, and 99% of things with certainty, I choose certainty.

-- 
see shy jo
